### PR TITLE
Block Directory: Update endpoints to remove `__experimental` namespace

### PR DESCRIPTION
**Pull request by ryelle**
## Description
This is in tandem with https://github.com/WordPress/wordpress-develop/pull/359, which adds the plugins & block directory endpoints into core under the `wp/v2` namespace. I also had to change how the assets are detected (JS vs CSS), since the regex wasn't matching files with version parameters (see https://github.com/WordPress/gutenberg/pull/23499).

I also changed the endpoints in gutenberg to use `wp/v2`, to preserve compatibility with sites not on trunk.

## How has this been tested?
I added a few blocks to a post, they added correctly. I removed some, and saved the post, the unused plugins were uninstalled correctly.

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
I think it might make sense to completely drop the tests now that the endpoint will be in core shortly. That matches what we did with the block types controller.

**Comment by TimothyBJacobs** [comment]
Test failure looks legit

**Comment by azaozz** [comment]
I'm probably missing something but how is "moving of end points" from Gutenberg to trunk supposed to work exactly? :) 

As far as I see there are couple of ways to do this:

1. Update the end points, change namespace from `__experimental` to `wp/v2`, etc. in both php and js (this is pretty much in this PR). Then ensure it's working as expected, merge the PR and publish the updated packages. Then make a core trunk patch and commit the updated packages together with moving the end points (php) from Gutenberg to trunk, and at the same time merge another PR that removes the end points from Gutenberg. For a short while the end points will be duplicate in core trunk and Gutenberg.

2. Update the end points, change namespace from `__experimental` to `wp/v2`, etc. in both php and js but do not merge the PR. Then make a patch and commit only the end points (php) to core trunk. Then update the PR, removing the end points. At this point the PR should be using trunk's end points. Then, if all is good, merge the PR, release the new packages and update them in core trunk.

Seems 2 is the better way, and that's what is being done here. Then seems that for this PR to move forward all of the php changes have to be committer to trunk first. Then the js changes can be merged.

**Comment by TimothyBJacobs** [comment]
> I'm probably missing something but how is "moving of end points" from Gutenberg to trunk supposed to work exactly? :)

It's a mess :) I don't think we've done it since 5.0 and the process there was kinda messy IIRC as we started merging things into 5.0 in Core.

> For a short while the end points will be duplicate in core trunk and Gutenberg.

> Then update the PR, removing the end points from Gutenberg. At this point the PR should be using trunk's end points.

I've been told, and don't really like, that the Gutenberg plugin needs to continue to work for 5.3 and 5.4 users. So we need to duplicate the endpoints in Gutenberg and in WordPress. The endpoints won't load in Gutenberg if the user is running trunk since the class is only included if it doesn't already exist ( see lib/load.php ).

I think this is really error prone, and not sustainable for the REST API endpoints. We'd need to backport any changes to those endpoints from Trunk back to GB so things wouldn't be out of sync.

> Seems 2 is the better way, and that's what is being done here.

I prefer that as well.

**Comment by azaozz** [comment]
> I've been told, and don't really like, that the Gutenberg plugin needs to continue to work for 5.3 and 5.4 users.
>
> I think this is really error prone, and not sustainable for the REST API endpoints. We'd need to backport any changes to those endpoints from Trunk back to GB so things wouldn't be out of sync.

Ah, I see. Yeah, seems quite messy... It's not just the end point php classes, (eventually) other newer code will have to be backported to support older WP.

Then [in 2](https://github.com/WordPress/gutenberg/pull/23528#issuecomment-650617444) above instead of removing the end points from Gutenberg, they will have to be replaced with what's in core trunk before merging the PR. Still seems like the better way to do this.

File: packages/block-directory/src/store/controls.js
Comment (StevenDufresne) [inline packages/block-directory/src/store/controls.js:14]: Just noting that this will force urls to be absolute based on `getPath`'s implementation.

**Comment by StevenDufresne** [comment]
I think we also want to update:

https://github.com/WordPress/gutenberg/blob/a2ed419b1c89ca6e855292810a5eb99a40eae9d8/packages/e2e-tests/specs/editor/plugins/block-directory-add.test.js#L87

File: packages/block-directory/src/store/controls.js
Comment (StevenDufresne) [reply packages/block-directory/src/store/controls.js:14]: We'll need to add the protocol in the End to End test:

https://github.com/WordPress/gutenberg/blob/a2ed419b1c89ca6e855292810a5eb99a40eae9d8/packages/e2e-tests/specs/editor/plugins/block-directory-add.test.js#L37

**Comment by ellatrix** [comment]
Didn't we agree to keep this endpoint experimental?

**Comment by TimothyBJacobs** [comment]
> Didn't we agree to keep this endpoint experimental?

Not to my knowledge. https://github.com/WordPress/wordpress-develop/pull/359#issuecomment-650253724

**Comment by ellatrix** [comment]
Ok, if it's ready then 👍

File: packages/block-directory/src/store/controls.js
Comment (ryelle) [reply packages/block-directory/src/store/controls.js:14]: It should always be a full URL from the API, once https://github.com/WordPress/gutenberg/pull/23499 is merged. This should be fixed in 8274c1c814b6841bfebb2b635409337516f4be95

**Comment by ryelle** [comment]
> I think it might make sense to completely drop the tests now that the endpoint will be in core shortly. That matches what we did with the block types controller.

Sure, I've removed them in 477d38e32fa4d27d96ee4d4c97e38075d719015a

> Then in 2 above instead of removing the end points from Gutenberg, they will have to be replaced with what's in core trunk before merging the PR. Still seems like the better way to do this.

That's what is happening here (combined with https://github.com/WordPress/gutenberg/pull/23499). The endpoints here will be the same as what are in trunk, and they all now live at `wp/v2/plugins` etc. This will keep compat with WP 5.4 when 5.5 comes out — 5.5 will load these endpoints from core, while 5.4 will load them from the plugin. It's not ideal to duplicate, but we can remove them when WP 5.6 comes out (added a note about that b537c3fdb84431965a26f0d295b48ab2de0fd1ca).

**Comment by TimothyBJacobs** [comment]
Do we want to switch out to the `userCan` API at this point?

**Comment by ryelle** [comment]
> Do we want to switch out to the userCan API at this point?

I thought that would be a good idea for a separate PR, so that this one is just about the endpoint changes.

**Review by TimothyBJacobs — APPROVED** [review]

**Comment by TimothyBJacobs** [comment]
> I thought that would be a good idea for a separate PR, so that this one is just about the endpoint changes.

Sounds good!

**Comment by youknowriad** [comment]
Merging this to fix the tests on master.

**Pull request merged**
